src/montgomery.rs

```diff
@@ -342,6 +370,9 @@ impl<'a, 'b> Mul<&'b Scalar> for &'a MontgomeryPoint {
             W: FieldElement::ONE,
         };
 
+        // NOTE: The below swap-double-add routine skips the first iteration, i.e., it assumes the
+        // MSB of `scalar` is 0. This is allowed, since it follows from Scalar invariant #1.
+
```

Comment:
I noticed that this was the case because the following test fails.
```rust
let a_bytes = [0xff; 32];
let s1 = Scalar::from_bytes_mod_order(a_bytes);
let s2 = Scalar { bytes: a_bytes };
assert_eq!(
    s1 * constants::X25519_BASEPOINT,
    s2 * constants::X25519_BASEPOINT
);
```
That's kinda surprising. There's no NAF or any fancy arithmetic necessary for the Montgomery ladder. Anyways, neither here nor there, because it works for all scalars < 2^255, which is all we care about.

Reply:
I mean, that's expected for me. One scalar is reduced, the other is not, but `Scalar { bytes: a_bytes }` would not be valid under `Scalar`'s (new) invariant and can't be constructed outside the crate due to field visibility.

sidebar: I'm curious if `Scalar` could be made into a newtype for `ScalarImpl` now.